source/io/Stream.cpp

```diff
@@ -13,132 +13,172 @@
+            }
 
-    impl->stream->seekg(currentPosition);
+            aws_input_stream_vtable InputStream::s_vtable = {
+                InputStream::s_Seek,
+                InputStream::s_Read,
+                InputStream::s_GetStatus,
+                InputStream::s_GetLength,
+                InputStream::s_Destroy,
+            };
 
-    return AWS_OP_SUCCESS;
-}
+            InputStream::InputStream(Aws::Crt::Allocator *allocator)
+            {
+                m_allocator = allocator;
+                m_underlying_stream =
+                    static_cast<aws_input_stream *>(aws_mem_calloc(m_allocator, 1, sizeof(aws_input_stream)));
 
-static void s_aws_input_stream_cpp_destroy(struct aws_input_stream *stream)
-{
-    aws_input_stream_cpp_impl *impl = static_cast<aws_input_stream_cpp_impl *>(stream->impl);
-    impl->stream = nullptr;
-    aws_mem_release(stream->allocator, stream);
-}
+                if (!m_underlying_stream)
+                {
+                    m_good = false;
+                    return;
+                }
 
-static struct aws_input_stream_vtable s_aws_input_stream_cpp_vtable = {s_aws_input_stream_cpp_seek,
-                                                                       s_aws_input_stream_cpp_read,
-                                                                       s_aws_input_stream_cpp_get_status,
-                                                                       s_aws_input_stream_cpp_get_length,
-                                                                       s_aws_input_stream_cpp_destroy};
+                m_good = true;
+                m_underlying_stream->impl = this;
+                m_underlying_stream->allocator = m_allocator;
+                m_underlying_stream->vtable = &s_vtable;
+            }
 
-namespace Aws
-{
-    namespace Crt
-    {
-        namespace Io
-        {
-            struct aws_input_stream *AwsInputStreamNewCpp(
-                const std::shared_ptr<Aws::Crt::Io::IStream> &stream,
+            StdIOStreamInputStream::StdIOStreamInputStream(
+                std::shared_ptr<Aws::Crt::Io::IStream> stream,
                 Aws::Crt::Allocator *allocator) noexcept
+                : InputStream(allocator), m_stream(std::move(stream))
             {
-                struct aws_input_stream *input_stream = NULL;
-                struct aws_input_stream_cpp_impl *impl = NULL;
-
-                aws_mem_acquire_many(
-                    allocator,
-                    2,
-                    &input_stream,
-                    sizeof(struct aws_input_stream),
-                    &impl,
-                    sizeof(struct aws_input_stream_cpp_impl));
-
-                if (!input_stream)
+            }
+
+            bool StdIOStreamInputStream::Read(ByteBuf &buffer) noexcept
+            {
+                m_stream->read(reinterpret_cast<char *>(buffer.buffer + buffer.len), buffer.capacity - buffer.len);
```

Comment:
I think you want to be using [readsome()](https://en.cppreference.com/w/cpp/io/basic_istream/readsome):
> Extracts up to count immediately available characters from the input stream

instead of [read()](https://en.cppreference.com/w/cpp/io/basic_istream/read):
>Characters are extracted and stored until any of the following conditions occurs:
>  * count characters were extracted and stored 
>  * end of file condition occurs on the input sequence (in which case, setstate(failbit|eofbit) is called). The number of successfully extracted characters can be queried using gcount(). 

also, readsome() returns the number of characters read, you don't need to call `gcount()` afterwards

Reply:
sorry readsome is a giant pile of worthless shit, Have to go back to read()